### PR TITLE
functionality upgrade

### PR DESCRIPTION
**Pull request by stefanbode**
## Description:
- added shutterxxx**0** command to change all shutters in parallel. Similar to power0 command
- added flexible ShutterMotorstop time to control the time between one relay OFF and the next ON. Now non blocking anymore other operations
- fix recent bug "shuttermode x" not working anymore
- changed autoconfig behavior to avoid SHORT CIRCUIT on missing interlock setting or wrong shuttermode
**Related issue (if applicable):** fixes #<Tasmota issue number goes here>

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [x] The code change is tested and works with Tasmota core ESP32 V.2.0.5
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

### CODE REVIEW

**Comment by arendst** [comment]
I'm about to release v12.3 today/tomorrow.

Do you trust your PR to be production proof? If so I'll merge before release.

**Comment by stefanbode** [comment]
Yes I trust. But let me fix the regression I see right now with timing. Let you know when everything is fine.

**Comment by stefanbode** [comment]
Merge is fine now. Also the last fixed issue was a minor one only happen if hammering the buttons and have a high motorstoptime. So quite rare anyhow but nobody likes a crash/reboot.

File: tasmota/include/tasmota_types.h [hunk 1]
```diff
@@ -730,8 +730,8 @@ typedef struct {
 #endif  // ESP32
   uint16_t      artnet_universe;           // 734
   uint16_t      modbus_sbaudrate;          // 736
-
-  uint8_t       free_esp32_738[5];         // 738
+  uint16_t      shutter_motorstop;         // 738
+  uint8_t       free_esp32_73A[3];         // 73A
```
Comment (arendst) [inline tasmota/include/tasmota_types.h:734]: By choosing this Setting location you are using part of the Settings area once used by other released Settings.

You can still use this BUT you will need to add value reset code to make the variable a default value.

Comment (stefanbode) [reply tasmota/include/tasmota_types.h:734]: The last fix reset anything above 64000. I will submit an additional change to limit it to 5000. this is physically the longest time I can imagine you should wait before moving the shutter again.

**Comment by purche42** [comment]
Hi Theo, hi Stefan,

I'm using Tasmota-ir for SOMFY RTS controlled shutters - besides many - many other Tasmotas in my home automation (garage door, smart couch ;), RF-Bridge etc.) .  

For my small Somfy RTS project on an ESP32 I can use as many virtual shutters I have in my home (10) . There is a limitation of four shutters which prevents the full usage of the great shutter integration. 
I have seen some discussion in 2020 around  MCP23xxx usage / raising the limit etc.  I also have seen increasing number of (virtual)relays and (virtual)buttons in latest versions. ESP32 have enough GPIOs for ~10 shutters which I also need for shutter integration but not for virtual - radio controlled devices. 

Are there any plans to change limit ? (Probably on ESP32 if there are memory constraints on other devices.)

Thanks for all your work - It makes my life easier and cloudless 

Peter

**Comment by purche42** [comment]
Hi Stephan,
maybe Theo has any plans to change the limits  ... Not super important  - but it would make the shutter integration much more useful.

Thanks BTW for the fast response

**Comment by arendst** [comment]
Shutters are Stefan (Bode) s lovechild.

Adding support for more would need major changes to current settings and at best would need a rewrite for ESP32. Considering the current implemntation I don't think this will happen soon.

So for the short term just add ESP32 for supporting more shutters. Adds redundancy too ;-)

**Comment by purche42** [comment]
Hi Theo,

thank you - I understand now that this is not an easy change. This works fine with additional logic in HASS and custom integration.  Tasmota would be the more elegant solution - as always :).
Also big thank for all you have done to make my smart home even smarter. 

Peter

**Comment by stefanbode** [comment]
Theo is right. Even if it is not a big deal with compiler options to change the shutter driver to more than 4 shutters I still have the dependency to store the configuration and latest values in a common predefined area. This is not flexible right now. Data is at predefined places in memory and must be reserved during developing a driver. Maybe if we can get a dynamic storage on ESP32 I can map it with less effort. As said @arendst , the static storage is the limitation.

**Comment by arendst** [comment]
@stefanbode Stefan, if you feel up to you can do the following:

- I've designed a way to extent settings storage using the filesystem. All our ESP32 environments have some kind of filesystem allowing to store settings as a file. This has already been exploited by four drivers. Have a look at ``xdrv_03_esp32_energy.ino``, ``xdrv_86_esp32_sonoff_spm.ino``, ``xdrv_87_esp32_sonoff_tm1621.ino`` and as a blueprint ``xdrv_122_file_settings_demo.ino``.  Search for keyword ``.drvset``.
- Using the above examples I suggest to add an ESP32 version of the shutter driver ``xdrv_27_esp32_shutter.ino`` keeping the current one solely for the ESP8266 and stable. Keep in mind future changes to your settings file layout like I did for the settings area to not burden the user with changes resulting in losing settings; use version control within the settings file which will proof very useful in your shutter writing code. Again, have a look at the examples given above.
- While developing consider removing the shutter button code. It is duplicating most of the current button driver while lacking support for virtual buttons. I think you didn't like this addition anyway.
- As file settings storage space is bound to filesystem sector size of 4k consider preparing for up to 32 relays while staying within this space allowing future drivers also to use the filesystem.

Take your time to implement and experiment before releasing this to the public.

Have fun.

**Comment by stefanbode** [comment]
Thanks Theo for the "onboarding". This is exactly what I need. I will check the methods there. Also duplicating the driver makes sense. Regarding the shutterbuttons: ok, not invented here but you take what you get. Maybe this is also an option to refactor it at least for the ESP32.

**Comment by stefanbode** [comment]
Making great progress. 16 shutter configuration seems working. Now I will refactoring the Shutterbuttons. I assume it will be also replace the ESP8266 one. Need to understand better how to get clear information of button press, release and hold. On shutters the expectation is that shutter starts moving with the button down event. The up event just define how to continue. Also a Double , tripple just changes the target position of an already running shutter.

**Comment by stefanbode** [comment]
@arendst : the driver demo file was the most useful one to get straight to a solution.

**Comment by stefanbode** [comment]
@arendst : I'm now fine with the refactoring. https://github.com/stefanbode/Sonoff-Tasmota/tree/shutter_ESP32
The only thing I just can "comment out" is the STATUS13 message. I have no idea how to fix this with the different approach.
ShutterButton removed all of the duplicate definition. Now much better in line with general buttons and the definitions.

**Comment by arendst** [comment]
Noticed your yell on discord. Will come back to it asap today.

**Comment by arendst** [comment]
If I understand correctly this might solve your issue:

I expect you make the "old" driver only available for ESP8266 and the "new" driver only for ESP32 like I did with the ``xdrv_03_energy.ino`` driver.

Change current code in ``support_command.ino`` to:
```
/*
#ifdef USE_SHUTTER
  if (Settings->flag3.shutter_mode) {
    if ((0 == payload) || (13 == payload)) {
      Response_P(PSTR("{\"" D_CMND_STATUS D_STATUS13_SHUTTER "\":{"));
      for (uint32_t i = 0; i < MAX_SHUTTERS; i++) {
        if (0 == Settings->shutter_startrelay[i]) { break; }
        if (i > 0) { ResponseAppend_P(PSTR(",")); }
        ResponseAppend_P(PSTR("\"" D_STATUS13_SHUTTER "%d\":{\"Relay1\":%d,\"Relay2\":%d,\"Open\":%d,\"Close\":%d,"
                                   "\"50perc\":%d,\"Delay\":%d,\"Opt\":\"%s\","
                                   "\"Calib\":[%d,%d,%d,%d,%d],"
                                   "\"Mode\":\"%d\"}"),
                                   i, Settings->shutter_startrelay[i], Settings->shutter_startrelay[i] +1, Settings->shutter_opentime[i], Settings->shutter_closetime[i],
                                   Settings->shutter_set50percent[i], Settings->shutter_motordelay[i], GetBinary8(Settings->shutter_options[i], 4).c_str(),
                                   Settings->shuttercoeff[0][i], Settings->shuttercoeff[1][i], Settings->shuttercoeff[2][i], Settings->shuttercoeff[3][i], Settings->shuttercoeff[4][i],
                                   Settings->shutter_mode);
      }
      ResponseJsonEndEnd();
      CmndStatusResponse(13);
    }
  }
#endif
*/
#ifdef USE_SHUTTER
  if ((0 == payload) || (13 == payload)) {
    if (ShutterStatus()) { CmndStatusResponse(13); }
  }
#endif
```

Add the function ``ShutterStatus()`` to both shutter drivers. The "old" driver should look like this:
```
bool ShutterStatus(void) {
  if (Settings->flag3.shutter_mode) {  // SetOption80  - (Shutter) Enable shutter support (1)
    Response_P(PSTR("{\"" D_CMND_STATUS D_STATUS13_SHUTTER "\":{"));
    for (uint32_t i = 0; i < MAX_SHUTTERS; i++) {
      if (0 == Settings->shutter_startrelay[i]) { break; }
      if (i > 0) { ResponseAppend_P(PSTR(",")); }
      ResponseAppend_P(PSTR("\"" D_STATUS13_SHUTTER "%d\":{\"Relay1\":%d,\"Relay2\":%d,\"Open\":%d,\"Close\":%d,"
                                  "\"50perc\":%d,\"Delay\":%d,\"Opt\":\"%s\","
                                  "\"Calib\":[%d,%d,%d,%d,%d],"
                                  "\"Mode\":\"%d\"}"),
                                  i, Settings->shutter_startrelay[i], Settings->shutter_startrelay[i] +1, Settings->shutter_opentime[i], Settings->shutter_closetime[i],
                                  Settings->shutter_set50percent[i], Settings->shutter_motordelay[i], GetBinary8(Settings->shutter_options[i], 4).c_str(),
                                  Settings->shuttercoeff[0][i], Settings->shuttercoeff[1][i], Settings->shuttercoeff[2][i], Settings->shuttercoeff[3][i], Settings->shuttercoeff[4][i],
                                  Settings->shutter_mode);
    }
    ResponseJsonEndEnd();
    return true;
  }
  return false;
}
```

The ESP32 driver should have a similar approach using your new setting values.

I hope you keep SO80 for enabling shutters like you do in the ESP8266 driver.

Does this answer your question?

**Comment by stefanbode** [comment]
Awesome. This make it.

**Comment by curzon01** [comment]
Great job, thx Stefan.
Not sure if you still working on refactoring: ESP32 drv GUI slider are currently not reflecting shutter position

**Comment by stefanbode** [comment]
Thanks for the hint. Major changes but cool

**Pull request merged**